[css-masonry] Fix intrinsic-sizing-cols-004-mix1/2

stwrt opened the pull request:
#### 3155098264ca1dfb212ba388794b7a4d8aa9d81a
<pre>
[css-masonry] Fix intrinsic-sizing-cols-004-mix1/2
<a href="https://bugs.webkit.org/show_bug.cgi?id=282340">https://bugs.webkit.org/show_bug.cgi?id=282340</a>
<a href="https://rdar.apple.com/problem/138927036">rdar://problem/138927036</a>

Reviewed by Sammy Gill.

During the stretch flexible tracks we need to take into account that indefinite items are also placed in each track.
This was being ignored before, which can cause differences in flex size computations.

Update mix1 test cases:

Test 2:
The &quot;5 5&quot; is set to a width of 2ch in the original test case.

Test 6:
An extra 5 was missing.

Test 9:
The last column should be ~3ch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1.html:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::IndefiniteSizeStrategy::accumulateFlexFractionMasonry const):
(WebCore::IndefiniteSizeStrategy::findUsedFlexFraction const):

Canonical link: <a href="https://commits.webkit.org/285993@main">https://commits.webkit.org/285993@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/da04d9ee3f1ff789baa906ee4a939f153e1e3115

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74406 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53835 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27217 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78787 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25644 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/76523 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62968 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1620 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/58474 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16800 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77473 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48639 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63993 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38882 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45645 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21488 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23977 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67035 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21835 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80306 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1723 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/995 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66762 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1871 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/64011 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66046 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16408 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9982 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8141 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1687 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/4475 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1716 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/1704 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1723 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
     return track.baseSize() / std::max<double>(1, flexFactor);
 }
 
+void IndefiniteSizeStrategy::accumulateFlexFractionMasonry(double& flexFraction, unsigned trackIndex, GridTrackSizingDirection outermostDirection, GridLayoutState& gridLayoutState) const
```
`outermostDirection` implies that we may recurse into subgrids but this does not seem to be the case for masonry so maybe just `direction`?

stwrt replied:
Subgrid support will be added in a future patch.

stwrt replied:
I converted the name to `direction` for now. Will need to change again when subgrid support is added.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+{
+    auto& allTracks = m_algorithm.tracks(outermostDirection);
+    auto trackLength = allTracks.size();
+    const Grid& grid = m_algorithm.grid();
```
```suggestion
    auto& grid = m_algorithm.grid();
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+void IndefiniteSizeStrategy::accumulateFlexFractionMasonry(double& flexFraction, unsigned trackIndex, GridTrackSizingDirection outermostDirection, GridLayoutState& gridLayoutState) const
+{
+    auto& allTracks = m_algorithm.tracks(outermostDirection);
+    auto trackLength = allTracks.size();
```
We don't seem to use `allTracks` so maybe:


```suggestion
    auto numTracks = m_algorithm.tracks(outermostDirection).size()
```

stwrt replied:
We only use `allTracks` in line 1362, so we can shorten this code by removing it.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+    auto trackLength = allTracks.size();
+    const Grid& grid = m_algorithm.grid();
+
+    for (auto track = 0u; track < trackLength; track++) {
```
`track` should be called `trackIndex`

stwrt replied:
We already have trackIndex defined above, so that does not work.

stwrt replied:
I renamed the function parameter, and updated `track` to `trackIndex`.

sammygill reviewed (commented):
Could you also include in your commit message an explanation of what you did to fix the fact that we were ignoring indefinite items before?

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
 }
 
+// FIXME: https://bugs.webkit.org/show_bug.cgi?id=282395
+void IndefiniteSizeStrategy::accumulateFlexFractionMasonry(double& flexFraction, unsigned columnIndex, GridTrackSizingDirection direction, GridLayoutState& gridLayoutState) const
```
How do we know this index is associated with the columns? AFAICT this code can run in both the column and rows direction. I think if we want to ditinguish between these two indices we may need to be a bit more descriptive in both their names as `trackIndex` probably won't suffice

stwrt replied:
You're right, it could be either columns or rows. I'll try to think of something more descriptive.

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+// FIXME: https://bugs.webkit.org/show_bug.cgi?id=282395
+void IndefiniteSizeStrategy::accumulateFlexFractionMasonry(double& flexFraction, unsigned columnIndex, GridTrackSizingDirection direction, GridLayoutState& gridLayoutState) const
+{
+    auto trackLength = m_algorithm.tracks(direction).size();
```
To be more consistent with the rest of the grid code: 
```suggestion
    auto numTracks = m_algorithm.tracks(direction).size();
```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,37 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+                    continue;
+
+                span = GridSpan::translatedDefiniteGridSpan(columnIndex, columnIndex + spanLength);
+            }
```
I feel like this reads slightly better but i'll leave it up to you to however you want to structure it:


```suggestion
            bool isIndefinitePlacedItem = GridPositionsResolver::resolveGridPositionsFromStyle(*m_algorithm.renderGrid(), *gridItem, direction).isIndefinite();
            if (!isIndefinitePlacedItem && span.startLine() != columnIndex) {
                continue;

            if (isIndefinitePlacedItem) {
              if (auto endLine = columnIndex + spanLength; endLine > trackLength)
                continue;
              span = GridSpan::translatedDefiniteGridSpan(columnIndex, columnIndex + spanLength);
            }

```

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1397,9 +1434,25 @@ double IndefiniteSizeStrategy::findUsedFlexFraction(Vector<unsigned>& flexibleSi
+            accumulateFlexFraction(flexFraction, iterator, direction, itemsSet, gridLayoutState);
+        }
+    } else {
+        Vector<RenderBox*> indefiniteItems;
```
Probably should use `Vector<WeakPtr<RenderBox>>`

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,42 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+            GridSpan span = m_algorithm.renderGrid()->gridSpanForGridItem(*gridItem, direction);
+
+            // Removing gutters from the max-content contribution of the item, so they are not taken into account in FindFrUnitSize().
+            LayoutUnit leftOverSpace = maxContentContributionForGridItem(*gridItem, gridLayoutState) - renderGrid()->guttersSize(direction, span.startLine(), span.integerSpan(), availableSpace());
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,42 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+            if (!itemsSet.add(*gridItem).isNewEntry)
+                continue;
+
+            GridSpan span = m_algorithm.renderGrid()->gridSpanForGridItem(*gridItem, direction);
```
auto

sammygill commented on Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
```diff
@@ -1355,6 +1356,42 @@ static inline double normalizedFlexFraction(const GridTrack& track)
+        span = GridSpan::translatedDefiniteGridSpan(flexTrackIndex, flexTrackIndex + spanLength);
+
+        // Removing gutters from the max-content contribution of the item, so they are not taken into account in FindFrUnitSize().
+        LayoutUnit leftOverSpace = maxContentContributionForGridItem(*gridItem, gridLayoutState) - renderGrid()->guttersSize(direction, span.startLine(), span.integerSpan(), availableSpace());
```
auto

stwrt committed:
[css-masonry] Fix intrinsic-sizing-cols-004-mix1/2
https://bugs.webkit.org/show_bug.cgi?id=282340
rdar://problem/138927036

Reviewed by Sammy Gill.

During the stretch flexible tracks we need to take into account that indefinite items are also placed in each track.
This was being ignored before, which can cause differences in flex size computations.

Update mix1 test cases:

Test 2:
The "5 5" is set to a width of 2ch in the original test case.

Test 6:
An extra 5 was missing.

Test 9:
The last column should be ~3ch.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-grid/masonry/tentative/intrinsic-sizing/masonry-intrinsic-sizing-cols-004-mix1.html:
* Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:
(WebCore::IndefiniteSizeStrategy::accumulateFlexFractionMasonry const):
(WebCore::IndefiniteSizeStrategy::findUsedFlexFraction const):

Canonical link: https://commits.webkit.org/285993@main
